### DISCUSSION TITLE
How can I match log from kernel buffer (dmesg)?

### DISCUSSION THREAD

**Discussion by GongT** [fail2ban/fail2ban#3892] [Q&A]
I have some firewall messages.

And I created `my-jail.conf` with `failregex = xxxx`.

Now how can I make it works?

**Comment by GongT** [comment]
I have tried `logpath = /proc/kmsg`. But this file never end, so it will not work.

**Reply by sebres** [reply]
> But this file never end

What do you mean with "never end"? 
- too long? - it is not a problem (fail2ban will scan only starting from last know position or else from `now - findtime`). 
- too wordy? - it depends on how good is your regex (if it is anchored and without catch-alls it'd be fast enough to process everything). 
- continuously grows? - well it is the matter of log-files... but there is logrotate that can help to hold them "shorter".

Besides `logpath` you have additionally to set `backend = auto`, to ensure the jail will monitor file and not systemd journal (in case your default backend is systemd).

**Comment by sebres** [comment]
The question "how can I make it works?" is not precise enough... What exactly is your problem?

If it is about proper failregex, see https://github.com/fail2ban/fail2ban/issues/1945#issuecomment-362247222
Otherwise please be a bit more precise with the description, we cannot look into your head.

**Reply by GongT** [reply]
My main problem is to read and matching kernel buffer, like what `dmesg` command do, no matter how.

the kernel buffer is not a real file on disk (`/var/log/messages` is not a "Linux" feature, so it do not exists on some system, e.g. on my Fedora 41 box).

`/proc/kmsg` file is a kernel feature (although it still able to opt-out). But when reading it, `while read()` loop will never break, it seems `fail2ban` not support this.

**Reply by sebres** [reply]
> But when reading it, `while read()` loop will never break

How so? What would happen when there is no messages anymore? Which backend did you try, polling or pyinotify?

In any case, aren't you able to redirect expected messages by (see /etc/rsyslog.conf), for example:
```
## by facility:
kern.*                          -/var/log/kern.log
## by some unique prefix:
:msg,startswith,"my_unique_prefix" /var/log/mylog.log
```
Or you'd probably monitor systemd journal instead (backend = `systemd` and corresponding `journalmatch`).
